ui/lib/api/core/core.pb.ts

```diff
@@ -132,6 +87,16 @@ export type GetObjectResponse = {
   object?: Gitops_coreV1Types.Object
 }
 
+export type ListObjectsRequest = {
+  namespace?: string
+  kind?: string
```

Comment:
I know, this file is auto-generated and see that all fields are generated as optional. So it's just a fantasy, but it would be nice if we could distinguish between non-optional and optional fields in requests in generated code.

For example, when listing objects for `ListObjectsRequest`, there is a kind lookup performed, and if it is not value/known, an error is returned. So, it would be nice to be able to generate kind as a non-optional value.

Or is it the Golang way: "everything can have a zeroed value", just throw an error if the value is not provided?

Reply:
It's actually not go, but the protobuf way - protobuf2 had a distinction between mandatory and optional fields, but protobuf3 removed it. The reason has to do with compatibility between versions of the protobuf schema.

In proto2, if the version of the client and the server are different, then one end might think a field is mandatory and the other end might not know about the field, and then they just refuse to talk to each other. Every time I've experienced this, it was because something that was once mandatory was no longer needed, but deploying that change would make the services no longer talk, and then I'd spend the rest of the day carefully coordinating multi-step deployments.

In proto3, every field is optional, every empty field is just even included in the message (so null, empty and missing are all the same thing), and every unknown field is just ignored. So as long as you don't re-use the numbers for different purposes between different versions, it will always work to parse a message. It might not be a useful result any more if the API has changed enough, but it will work.

This doesn't really apply to our use because we ship and deploy the server and client together, so they should always have the same version. But that's why.

Reply:
Thanks for the detailed explanation, good to know!